src/core/utils/flash.cpp

```diff
@@ -39,44 +39,281 @@
+    };
+
+public:
+    inline void Load(otInstance *aInstance, uint8_t aSwapIndex, uint32_t aOffset)
+    {
+        otPlatFlashRead(aInstance, aSwapIndex, aOffset, this, sizeof(*this));
+    }
+
+    inline void Save(otInstance *aInstance, uint8_t aSwapIndex, uint32_t aOffset) const
+    {
+        otPlatFlashWrite(aInstance, aSwapIndex, aOffset, this, sizeof(*this));
+    }
+
+    inline uint16_t ReadData(otInstance *aInstance,
+                             uint8_t     aSwapIndex,
+                             uint32_t    aOffset,
+                             uint8_t *   aValue,
+                             uint16_t    aLength)
+    {
+        if (aLength > GetLength())
+        {
+            aLength = GetLength();
+        }
+
+        otPlatFlashRead(aInstance, aSwapIndex, aOffset + sizeof(*this), aValue, aLength);
+        return aLength;
+    }
+
+    uint16_t GetKey(void) const { return mKey; }
+
+    uint16_t GetLength(void) const { return mLength; }
+
+    uint16_t GetDataSize(void) const
+    {
+        uint16_t wordMask = GetAlignmentMask();
+        return (mLength + wordMask) & ~wordMask;
+    }
+
+    static uint16_t GetAlignmentMask(uint8_t aFormat)
+    {
+        return (aFormat == kFormatV1) ? (kFlashWordSizeV1 - 1) : (kFlashWordSizeV2 - 1);
+    }
+
+    uint16_t GetSize(void) const { return sizeof(*this) + GetDataSize(); }
+
+    bool IsValid(void) const { return ((mFlags & (kFlagAddComplete | kFlagDelete)) == kFlagDelete); }
+
+    void SetDeleted(void)
+    {
+        memset(this, 0xff, sizeof(*this));
+        mFlags &= ~kFlagDelete;
+    }
+
+    bool IsFirst(void) const { return (mFlags & kFlagFirst) == 0; }
+
+    bool IsFree(void) const { return !IsAddBeginSet(); }
+
+protected:
+    bool IsIntegrityOk(void) const
+    {
+        OT_ASSERT(IsAddBeginSet());
+        return IsAddCompleteSet();
+        // TODO: validate GetFormat() == 0 && mReserved == 0xffff;
+    }
+
+    uint8_t GetFormat() const { return kFormatTop - (mFlags >> kFormatShift); }
+
+private:
+    uint16_t GetAlignmentMask(void) const { return GetAlignmentMask(GetFormat()); }
+
+    bool IsAddBeginSet(void) const { return (mFlags & kFlagAddBegin) == 0; }
+    bool IsAddCompleteSet(void) const { return (mFlags & kFlagAddComplete) == 0; }
+
+    enum
+    {
+        kFlagAddBegin    = 1 << 0, ///< 0 indicates record write has started, 1 otherwise.
+        kFlagAddComplete = 1 << 1, ///< 0 indicates record write has completed, 1 otherwise.
+        kFlagDelete      = 1 << 2, ///< 0 indicates record was deleted, 1 otherwise.
+        kFlagFirst       = 1 << 3, ///< 0 indicates first record for key, 1 otherwise.
+
+        /* The record's format version is in the upper nibble of mFlags */
+        kFormatTop   = 0xf,
+        kFormatShift = 12,
+
+        /* Initial value of flags: kFlagAddBegin, kFlagAddComplete and format version */
+        kFlagsInit = ((kFormatTop - kFormatV2) << kFormatShift) | (0xfff & ~(kFlagAddBegin | kFlagAddComplete)),
```

Comment:
It seems we are not attempting to maintain backward compatibility in that the V1 flash implementation will not be able to read the V2 flash implementation records. If that is the case, do we care to maintain the begin/complete flags as well?

Reply:
Actually, we don't need to maintain both flags. This is to simplify some methods which would depend on the format version. For example, see the `IsFree` and `IsIntegrityOk` methods. I think the code is a lot more complex/uglier than it should be, because of legacy compatibility.

Reply:
I understand the need to understand the old format to support forward migration. However, if we are not looking to support backward migration (i.e. writing old format), then I think we should do whatever is most clean going forward.

Reply:
Yes, I was hoping to do some "cleanup" after the main functions were implemented.